File: mask-operations/contrast_n_brightness.py
#-----------------------------------------------------------------------------
#
# Example for contrast and brightness changing in Python.
#
#-----------------------------------------------------------------------------
#
# Hand-written version of cv2.convertScaleAbs() function.
#
#------------------------------------------------------------------------------

import cv2
import numpy as np

def saturate_cast(value, dtype):
    return np.clip(value, np.iinfo(dtype).min, np.iinfo(dtype).max).astype(dtype)

alpha = 2
beta = -50
src = cv2.imread("../img.jpg")
img = cv2.imread("../img.jpg")

# Change all pixels using the formula `A*pixel + B`.
for i in range(img.shape[0]):
    for j in range(img.shape[1]):
        for c in range(img.shape[2]):
            img[i, j, c] = saturate_cast(
                alpha * img[i, j, c] + beta, dtype=img.dtype
            )

cv2.namedWindow("new")
cv2.imshow("new", img)
cv2.namedWindow("src")
cv2.imshow("src", src)

cv2.waitKey(0)
cv2.destroyAllWindows()

File: mask-operations/contrast_n_brightness.cpp
//-----------------------------------------------------------------------------
//
// Example for contrast and brightness changing in C++.
//
//-----------------------------------------------------------------------------
//
// Hand-written version of cv::convertScaleAbs() function.
//
//------------------------------------------------------------------------------

#include <opencv2/opencv.hpp>

int main() {
    float alpha = 2, beta = -50;
    cv::Mat src, img;

    src = cv::imread("../img.jpg");
    img = cv::imread("../img.jpg");

    // Change all cells for formula `A*pixel + B`.
    for (int i = 0; i != img.rows; ++i) {
        for (int j = 0; j != img.cols; ++j) {
            for (int c = 0; c != img.channels(); ++c) {
                uchar& cellRef = img.at<cv::Vec3b>(i, j)[c];
                cellRef = cv::saturate_cast<uchar>(cellRef * alpha + beta);
            }
        }
    }

    cv::namedWindow("new");
    cv::imshow("new", img);
    cv::namedWindow("src");
    cv::imshow("src", src);

    cv::waitKey();
    cv::destroyAllWindows();
}
